example/grammar/magnet.cpp

```diff
@@ -0,0 +1,496 @@
+//
+// Copyright (c) 2022 alandefreitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0.
+// https://www.boost.org/LICENSE_1_0.txt
+//
+
+#include <boost/url/url_view.hpp>
+#include <boost/url/rfc/absolute_uri_rule.hpp>
+#include <boost/url/grammar/charset.hpp>
+#include <iostream>
+
+using namespace boost::urls;
+
+// A small filter and transform view
+template <class C, class V, class P, class T>
+class filter_view
+{
+    C c_;
+    P p_;
+    T t_;
+
+    using value_type = V;
+
+public:
+    class iterator
+    {
+        friend filter_view;
+
+        using base_iter = typename C::iterator;
+
+        base_iter b_;
+        filter_view& v_;
+
+        void
+        increment_false()
+        {
+            auto e = v_.c_.end();
+            while (b_ != e && !v_.p_(*b_))
+                ++b_;
+        }
+
+        iterator(
+            filter_view& v,
+            base_iter it)
+            : b_(it)
+            , v_(v)
+        {
+            increment_false();
+        }
+
+    public:
+        iterator&
+        operator++() noexcept
+        {
+            ++b_;
+            increment_false();
+            return *this;
+        }
+
+        iterator
+        operator++(int) noexcept
+        {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        value_type
+        operator*() const
+        {
+            return v_.t_(*b_);
+        }
+
+        friend
+        bool
+        operator==(
+            iterator const& a,
+            iterator const& b) noexcept
+        {
+            return a.b_ == b.b_;
+        }
+
+        friend
+        bool
+        operator!=(
+            iterator const& a,
+            iterator const& b) noexcept
+        {
+            return a.b_ != b.b_;
+        }
+    };
+
+    filter_view(
+        C const& base,
+        P pred = {},
+        T transform = {})
+        : c_(base)
+        , p_(pred)
+        , t_(transform) {};
+
+    iterator begin()
+    {
+        return {*this, c_.begin()};
+    }
+
+    iterator end()
+    {
+        return {*this, c_.end()};
+    }
+};
+
+struct magnet_link_rule_t;
+
+// A new url type for magnet links
+class magnet_link_view
+{
+    url_view u_;
+
+    struct is_exact_topic {
+        bool operator()(query_param_encoded_view p)
+        {
+            if (p.key == "xt")
```

Comment:
A comment would be nice here explaining that it is using the lazy op== (or is it?)

Reply:
I changed it to `query_param_view` so we can use the proper `pct_encoded_view` and `op==`. `query_param_view` is also the right thing to do because these things are allowed to be encoded.

I ended up using `query_param_encoded_view` because things were becoming inconvenient with `pct_encoded_view` since it doesn't have some functions like `starts_with`, etc...